src/commons/utils/JavaHelper.ts

```diff
@@ -0,0 +1,145 @@
+import setupJVM, { parseBin } from 'java-slang/dist/jvm';
+import { createModuleProxy, loadCachedFiles } from 'java-slang/dist/jvm/utils/integration';
+import { Context } from 'js-slang';
+import { initModuleContext, loadModuleBundle } from 'js-slang/dist/modules/loader/moduleLoader';
+
+import DisplayBufferService from './DisplayBufferService';
+
+const CDN = 'https://source-academy.github.io/modules/java/java-packages/src/';
+
+export async function javaRun(javaCode: string, context: Context) {
+  let compiled = {};
+
+  let files = {};
+  let buffer: string[] = [];
+
+  const readClassFiles = (path: string) => {
+    let item = files[path as keyof typeof files];
+
+    // not found: attempt to fetch from CDN
+    if (!item && path) {
+      const splits = path.split('/');
+      splits.pop(); // classname.class
+      const pkg = splits.join('_');
+
+      const request = new XMLHttpRequest();
+      request.open('GET', `${CDN}${pkg}.json`, false);
+      request.send(null);
+      if (request.status !== 200) {
+        throw new Error('File not found: ' + path);
+      }
+      const json = JSON.parse(request.responseText);
+      // we might want to cache the files in IndexedDB here
+      files = { ...files, ...json };
+
+      if (!files[path as keyof typeof files]) {
+        throw new Error('File not found: ' + path);
+      }
+
+      item = files[path as keyof typeof files];
+    }
+
+    // convert base64 to classfile object
+    const binaryString = atob(item);
+    const bytes = new Uint8Array(binaryString.length);
+    for (let i = 0; i < binaryString.length; i++) {
+      bytes[i] = binaryString.charCodeAt(i);
+    }
+    return parseBin(new DataView(bytes.buffer));
+  };
+  const loadNatives = (path: string) => {
+    // dynamic load modules
+    if (path.startsWith('modules')) {
+      const module = path.split('/')[1] as string;
+      initModuleContext(module, context, true);
+      const moduleFuncs = loadModuleBundle(module, context);
+      const { proxy } = createModuleProxy(module, moduleFuncs);
+      return Promise.resolve({ default: proxy });
+    }
+
+    return import(`java-slang/dist/jvm/stdlib/${path}.js`).then(m => {
+      return m;
+    });
+  };
+  const stdout = (str: string) => {
+    // java flush pushes a single newline to buffer
+    if (str === '\n') {
```

Comment:
What happens when you intentionally want to `System.out.println('');`?

Reply:
This is kind of a problem here. If we ignore this and always print whatever comes to stdout, we would get extra newlines. If we do it this way, `System.out.println('');` doesnt print a new line.

![image](https://github.com/source-academy/frontend/assets/37036869/71a75965-b7f1-4a75-ad48-4493dfeeb6f8)
![image](https://github.com/source-academy/frontend/assets/37036869/3c026a62-475f-49a0-81a2-d1d06d377a6b)

Not sure which is better.

Reply:
Hmmm, I tried looking at [Java documentation](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/PrintStream.html#flush()) and StackOverflow, and the idea I'm getting is that there are two types of flushes:

* Implicit flush (done every time `\n` is seen)
* Explicit flush (done by calling `.flush()`)

And seems like `.flush()` does not print an additional newline?

Though the screenshots definitely show otherwise, so I'm not sure of the implementation detail; what exactly is causing this flush to happen for us?

Reply:
I am not sure how `println` is implemented, but as far as the standard library is concerned we only implement `FiledOutputStream::writeBytes`: [implementation](https://github.com/source-academy/java-slang/blob/main/src/jvm/stdlib/java/io/FileOutputStream.ts#L14)

`println` is not implemented by us, that the java standard library.

What I notice is when we `println("a")` we get `writeBytes("a")` and `writeBytes("\n")`. `DisplayBufferService` logs this as two separate lines rather than a single `a\n`

Reply:
![image](https://github.com/source-academy/frontend/assets/37036869/79106cf8-5e63-4dc0-bb17-41592d0c0c86)
Think I have a better solution, we just split newlines and push each line to `DisplayBufferService`.